### PR TITLE
e2e document part1

### PR DESCRIPTION
**Pull request by itsmekumari**
Hi @tivv @rmstar Added e2e test document part1. Please review.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,5 @@
-# ODP
\ No newline at end of file
+# ODP
```
Comment (rmstar) [inline README.md]: This is unrelated to ODP, please fix

File: README.md [hunk 2]
```diff
@@ -1 +1,5 @@
...
+
+1. How to run the test against (a) local sandbox (b) remote CDAP instance. REFER Runtest.md
```
Comment (rmstar) [inline README.md]: move the contents of Runtest.md here - no need for a separate Runtest.mdfile.
Add sections for debugging test failures and adding new tests in this file, and link to the relevant docs

File: Runtest.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# How to run the test against a) Local sandbox and b) Remote CDAP instance. 
+
+## Local sandbox: 
+
+Provide the local sandbox URL in ConnectionParameters.properties file.
```
Comment (rmstar) [inline Runtest.md]: Add link to ConnectionParameters.properties

File: Runtest.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
+# How to run the test against a) Local sandbox and b) Remote CDAP instance. 
+
+## Local sandbox: 
+
+Provide the local sandbox URL in ConnectionParameters.properties file. 
+
+In the ConnectionParameters.properties file, provide the local sandbox URL. The path of the property file is 
+cdap-e2e-tests>src>main>resource>ConnectionParameters.properties
```
Comment (rmstar) [inline Runtest.md]: Don't specify path like this - just link to the file.

File: Runtest.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+Property name and its value is CDFURL=http://localhost:11011/pipelines/ns/default/studio
+All the configuration details need to be present in this file.
```
Comment (rmstar) [inline Runtest.md]: Describe the config parameters in this file. What does the user need to change if they want to run the test against their sandbox?

File: Runtest.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+**Configuration Details required for Setup:**
+
+The cdap-e2e is a framework and google-cloud/src/e2e-test is a plugin specific repo, which contains a plugin specific test and step definition.
```
Comment (rmstar) [inline Runtest.md]: Don't refer to google-cloud repo here. The test framework should work with any plugin repo, not just google-cloud.

File: Runtest.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+**Test Runner:**
+
+We can run the feature file and execute from the Test runner. Open the Test runner file and give the feature file to be executed.
+The Test Runner file has glue and tags configurations that describe the location and path of the step definition file and tags which are defined in the feature file. It can be specified like below.
+glue = {"io.cdap.plugin.gcp.stepsdesign", "stepsdesign"},
+tags = {"@BQMT"},
+
+## Remote CDAP instance
+
+Provide the sandbox URL in ConnectionParameters.properties file.
```
Comment (rmstar) [inline Runtest.md]: sandbox->CDAP

File: Runtest.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+We can run the feature file and execute from the Test runner. Open the Test runner file and give the feature file to be executed.
+The Test Runner file has glue and tags configurations that describe the location and path of the step definition file and tags which are defined in the feature file. It can be specified like below.
+glue = {"io.cdap.plugin.gcp.stepsdesign", "stepsdesign"},
+tags = {"@BQMT"},
+
+## Remote CDAP instance
+
+Provide the sandbox URL in ConnectionParameters.properties file. 
+
+In the ConnectionParameters.properties file, provide the sandbox URL to access the Remote CDAP instance. The path of the property file is
```
Comment (rmstar) [inline Runtest.md]: provide the URL to access the remote CDAP instance

File: Runtest.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+The Test Runner file has glue and tags configurations that describe the location and path of the step definition file and tags which are defined in the feature file. It can be specified like below.
+glue = {"io.cdap.plugin.gcp.stepsdesign", "stepsdesign"},
+tags = {"@BQMT"},
+
+## Remote CDAP instance
+
+Provide the sandbox URL in ConnectionParameters.properties file. 
+
+In the ConnectionParameters.properties file, provide the sandbox URL to access the Remote CDAP instance. The path of the property file is 
+cdap-e2e-tests>src>main>resource>ConnectionParameters.properties
```
Comment (rmstar) [inline Runtest.md]: No need specify path, just link to the file.

File: Runtest.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+* cdap-e2e-tests
+* google-cloud
+
+_Clone the cdap-e2e-tests framework of develop branch, which contains the common helper classes:_
+
+1. Open IntelliJ IDEA tool.
+2. From the File menu select New and choose Project From Version Control option.
+3. Provide the Github URL in the URL and click on Clone to clone the project.
+
+_URL: https://github.com/cdapio/cdap-e2e-tests_
```
Comment (rmstar) [inline Runtest.md]: This is not necessary, can be removed

File: Runtest.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+1. Now clone the google-cloud and data-integration project. From the File menu select New and choose Project From Version Control option.
+2. Provide the Github URL in the URL. Click Clone to clone the project.
+_URL: https://github.com/data-integrations/google-cloud_
+3. Select Maven and verify if the e2e tests profile is selected. Click reload all maven projects to reload the dependencies.
+4. Edit the configurations for cloud service for Test runner to maintain the service account. The service account key has permission to access and connect to the GCP client, which is used for creation and validation.
+>GOOGLE_APPLICATION_CREDENTIALS=key [Path of JSON service key]
+5. To run the profile enter mvn verify -P e2e-tests command in terminal.
+6. After the build is successful, then the target folder will be created automatically.
+7. In the console we can see the below screenshot in execution from maven displaying the results.
```
Comment (rmstar) [inline Runtest.md]: "below screenshot" -> where is the screenshot

File: Runtest.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+_Please find the list of connectors along with their specific configuration details:-_
+
+1. GCS: When we have GCS as source, we have to provide the path of the file located in the GCS bucket which already exists. All the different file formats paths used for testing are configured here, it also includes the GCP credentials. 
+
+GCSbucket- Provide the bucket path
+For example, gs://<bucket>/path/to/directory/
+			 project-id= XXXXX
+
+### Add Plugin specific Actions and Locators:
```
Comment (rmstar) [inline Runtest.md]: This section should be moved to the doc that describes how to create new tests

File: Runtest.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+**Add Actions**
+1. It contains locator actions of what action needs to be performed on the respective field of the plugin in properties.
+2. It contains the Util classes example cdfHelper class which contains common methods related to CDF and can implement them in step definitions, and SeleniumHelper class which contains common java functions.
+3. After taking the latest pull every time, in the terminal enter maven clean install command to compile. Create a JAR and inject the dependencies in integration project.
+
+
+_Clone the google-cloud, data-integration framework which contains the plugin specific details:_
+
+1. Now clone the google-cloud and data-integration project. From the File menu select New and choose Project From Version Control option.
```
Comment (rmstar) [inline Runtest.md]: You can give google-cloud as an example, but this is not specific to google-cloud. You can refer to plugin repo with e2e test framework integration.

File: Runtest.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+2. Provide the Github URL in the URL. Click Clone to clone the project.
+_URL: https://github.com/data-integrations/google-cloud_
+3. Select Maven and verify if the e2e tests profile is selected. Click reload all maven projects to reload the dependencies.
```
Comment (rmstar) [inline Runtest.md]: Details of using intellij are not needed. Just specify the command to run the tests

File: README.md (hunk 1 above)
Comment (itsmekumari) [reply README.md]: Modified the text.

File: Runtest.md (hunk 11 above)
Comment (itsmekumari) [reply Runtest.md]: Removed the text.

File: Runtest.md (hunk 10 above)
Comment (itsmekumari) [reply Runtest.md]: Removed the mentioned text.

File: Runtest.md (hunk 14 above)
Comment (itsmekumari) [reply Runtest.md]: Details are removed.

File: README.md (hunk 2 above)
Comment (itsmekumari) [reply README.md]: Done.

File: Runtest.md (hunk 3 above)
Comment (itsmekumari) [reply Runtest.md]: Done.

File: Runtest.md (hunk 4 above)
Comment (itsmekumari) [reply Runtest.md]: Done.

File: Runtest.md (hunk 8 above)
Comment (itsmekumari) [reply Runtest.md]: Done.

File: Runtest.md (hunk 5 above)
Comment (itsmekumari) [reply Runtest.md]: Done.

File: Runtest.md (hunk 6 above)
Comment (itsmekumari) [reply Runtest.md]: Done.

File: Runtest.md (hunk 7 above)
Comment (itsmekumari) [reply Runtest.md]: Done.

Comment (itsmekumari) [reply Runtest.md]: @rmstar At multiple places we have sandbox, so need to replace it with CDAP?

File: Runtest.md (hunk 9 above)
Comment (itsmekumari) [reply Runtest.md]: Done.

File: Runtest.md (hunk 12 above)
Comment (itsmekumari) [reply Runtest.md]: Done.

File: Runtest.md (hunk 13 above)
Comment (itsmekumari) [reply Runtest.md]: Modified the text.

File: README.md [hunk 15]
```diff
@@ -1 +1,59 @@
...
+
+##1. How to run the test against a) Local sandbox and b) Remote CDAP instance.
+
+### Local sandbox:
+
+Provide the local sandbox URL in [ConnectionParameters.properties](src/main/resources/ConnectionParameters.properties) file.
+
+In the ConnectionParameters.properties file, provide the local sandbox URL in
+[ConnectionParameters.properties](src/main/resources/ConnectionParameters.properties)
+
```
Comment (itsankit-google) [inline README.md:11]: Line no 7 and Line no 9, 10 look pretty similar to me. Can we consolidate them?

File: README.md [hunk 16]
```diff
@@ -1 +1,59 @@
...
+All the configuration details need to be present in this file specific to test.
+Here configuration parameters are defined to use everytime wherever required. Changes are done directly here if need to change any particular value. Below parameter values need to be changed to run the test against different sandbox.
+
+Property name and its value is CDFURL=http://localhost:11011/pipelines/ns/default/studio
+Project-ID=cdf-athena<br>
+@TC-Demo-1_GCS=gs://cdf-athena-test/test.csv<br>
+@TC-Demo-2_GCS=gs://cdf-athena-test/test-DataTypes.csv<br>
+Data-Set=test_automation<br>
+tableDemo=DemoCheck1<br>
+
```
Comment (itsankit-google) [inline README.md]: I don't think anything other than cdfUrl is specific to sandbox. @rmstar what do you think?

File: README.md (hunk 15 above)
Comment (itsmekumari) [reply README.md:11]: Done.

File: README.md (hunk 16 above)
Comment (itsmekumari) [reply README.md]: Done from my side. @rmstar Please confirm your inputs.

Comment (rmstar) [reply README.md]: projectId config will need to be changed if the user wants to run the test on a different. They might not have access to cdf-athena.

File: README.md [hunk 17]
```diff
@@ -1 +1,31 @@
...
+**Configuration Details required for Setup:**
+
+1. Create a maven profile in the main project for which e2e tests are required, with a dependency of e2e test  framework artifact, along with your Runner file inside the failsafe plugin.
+
+NOTE: All the properties/ actions/ locators which are common to plugins, need to be or will be present in the framework code. Plugin specific properties/ actions/ locators should reside in the main project module (for eg: google-cloud).
+
+2. TestRunners and its step definition should also resides in the main project module. 
+
+
```
Comment (rmstar) [inline README.md]: nit: remove extra line

File: README.md [hunk 18]
```diff
@@ -1 +1,31 @@
...
+### a) Local CDAP sandbox:
+
+Provide the local CDAP sandbox URL in [connectionParameters.properties](src/main/resources/connectionParameters.properties) file.
+>cdfurl=http://localhost:11011/pipelines/ns/default/studio
+
+Other keys (like @TC-Demo-1_GCS, Data-Set, etc) can be modified as needed. <br>
+
+**Configuration Details required for Setup:**
+
+1. Create a maven profile in the main project for which e2e tests are required, with a dependency of e2e test  framework artifact, along with your Runner file inside the failsafe plugin.
```
Comment (rmstar) [inline README.md]: nit: remove extra space - "test  framework"

File: README.md [hunk 19]
```diff
@@ -1 +1,31 @@
...
+##1. How to run the e2e test against :
+
+### a) Local CDAP sandbox:
+
+Provide the local CDAP sandbox URL in [connectionParameters.properties](src/main/resources/connectionParameters.properties) file.
+>cdfurl=http://localhost:11011/pipelines/ns/default/studio
+
+Other keys (like @TC-Demo-1_GCS, Data-Set, etc) can be modified as needed. <br>
+
+**Configuration Details required for Setup:**
```
Comment (rmstar) [inline README.md]: This section describes how to add e2e tests in a plugin repo. I think it should be moved to Addnewplugin.md.

File: README.md [hunk 20]
```diff
@@ -1 +1,31 @@
-# ODP
\ No newline at end of file
+# E2E-Test Framework
+
+##1. How to run the e2e test against :
```
Comment (rmstar) [inline README.md:3]: please check if the section numbers are rendered correctly

File: README.md [hunk 21]
```diff
@@ -1 +1,31 @@
...
+**Configuration Details required for Setup:**
+
+1. Create a maven profile in the main project for which e2e tests are required, with a dependency of e2e test  framework artifact, along with your Runner file inside the failsafe plugin.
+
+NOTE: All the properties/ actions/ locators which are common to plugins, need to be or will be present in the framework code. Plugin specific properties/ actions/ locators should reside in the main project module (for eg: google-cloud).
+
+2. TestRunners and its step definition should also resides in the main project module. 
+
+
+3. The service account file has permission to access and connect to the GCP client, which is used for creation and validation.
```
Comment (rmstar) [inline README.md]: It's not clear what this means. Please rephrase.

File: README.md (hunk 17 above)
Comment (itsmekumari) [reply README.md]: Removed.

File: README.md (hunk 18 above)
Comment (itsmekumari) [reply README.md]: Removed.

File: README.md (hunk 21 above)
Comment (itsmekumari) [reply README.md]: Done.

File: README.md (hunk 16 above)
Comment (itsmekumari) [reply README.md]: Done.

File: README.md (hunk 20 above)
Comment (itsmekumari) [reply README.md:3]: Yes, verified.

File: README.md (hunk 19 above)
Comment (itsmekumari) [reply README.md]: Moved.

File: README.md [hunk 22]
```diff
@@ -1 +1,18 @@
...
+>cdfurl=http://localhost:11011/pipelines/ns/default/studio
+
+Other keys (like @TC-Demo-1_GCS, Data-Set, Project-ID etc.) can be modified as needed. <br>
+
+### b) Remote CDAP instance:
+
+Modify the key cdfurl in [connectionParameters.properties](src/main/resources/connectionParameters.properties) to connect to remote CDAP instance.
+>cdfurl=[http://*`ip address`*:*`port`*/pipelines/ns/default/studio]()
+
+##2. How to debug test failures. where logs and screenshots can be found. REFER [Debugtest.md](Debugtest.md)
```
Comment (rmstar) [inline README.md:17]: Add newline between sections, make sure this renders correctly.

File: README.md [hunk 23]
```diff
@@ -1 +1,18 @@
...
+# E2E-Test Framework
+
+##1. How to run the e2e test against :
+
+### a) Local CDAP sandbox:
+
+Provide the local CDAP sandbox URL in [connectionParameters.properties](src/main/resources/connectionParameters.properties) file.
+>cdfurl=http://localhost:11011/pipelines/ns/default/studio
+
+Other keys (like @TC-Demo-1_GCS, Data-Set, Project-ID etc.) can be modified as needed. <br>
```
Comment (rmstar) [inline README.md]: Looks like [@TC-Demo-1_GCS](https://github.com/cdapio/cdap-e2e-tests/blob/develop/src/main/resources/connectionParameters.properties#L7) is going to be removed - remove reference to it. Also, please add comments in connectionParameters.properties indicating what each property is used for.

File: README.md (hunk 22 above)
Comment (itsmekumari) [reply README.md:17]: Done.

File: README.md (hunk 23 above)
Comment (itsmekumari) [reply README.md]: Removed. And we will take this in a separate PR and add we have taken this in our backlog.

**Review by rmstar — APPROVED** [review]
please squash commits

**Review by itsankit-google — COMMENTED** [review]
Closing this as duplicate of #45
